### PR TITLE
collate and minify most scripts loaded by main.html

### PR DESCRIPTION
**Pull request by mykmelez**
Here's an actual implementation of the script collation demonstrated by #1297. Still a WIP, as I'd like to see if I can improve it a bit. But note that it already does pretty well on startup time tests:

|      Test     | Baseline Mean |   Mean   |   +/-  |    %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|-------:|-------:|--------------:|---------:|---------:|
| startupTime   |       1,334ms |  1,218ms | -117ms |  -8.75 |        BETTER |  1,171ms |  1,346ms | 
| totalSize     |      33,691kb | 33,557kb | -135kb |  -0.40 | INSIGNIFICANT | 31,999kb | 34,728kb | 
| domSize       |         208kb |    170kb |  -38kb | -18.46 | INSIGNIFICANT |    170kb |    170kb | 
| styleSize     |         433kb |    432kb |    0kb |  -0.09 |        BETTER |    432kb |    433kb | 
| jsObjectsSize |      17,004kb | 17,001kb |   -4kb |  -0.02 |        BETTER | 17,000kb | 17,005kb | 
| jsStringsSize |         635kb |    635kb |    0kb |  -0.02 | INSIGNIFICANT |    635kb |    635kb | 
| jsOtherSize   |      14,975kb | 14,889kb |  -86kb |  -0.58 | INSIGNIFICANT | 13,326kb | 16,059kb | 
| otherSize     |         436kb |    430kb |   -6kb |  -1.33 |        BETTER |    430kb |    430kb |

### CODE REVIEW

**Comment by marco-c** [comment]
The Travis build is failing because it can't download the SpiderMonkey shell (fixed in #1318)

**Comment by mykmelez** [comment]
A few scripts are harder to collate, because of dependencies or other issues. But this branch now collates most of them into *bld/main-all.js*. And it improves startup time for a complex midlet on desktop:

|      Test     | Baseline Mean |   Mean   |   +/-  |    %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|-------:|-------:|--------------:|---------:|---------:|
| startupTime   |       1,279ms |  1,167ms | -112ms |  -8.79 |        BETTER |  1,137ms |  1,217ms | 
| totalSize     |      26,821kb | 25,968kb | -854kb |  -3.18 |        BETTER | 25,084kb | 26,842kb | 
| domSize       |         187kb |    153kb |  -34kb | -17.94 |        BETTER |    153kb |    154kb | 
| styleSize     |         433kb |    432kb |    0kb |  -0.02 | INSIGNIFICANT |    432kb |    433kb | 
| jsObjectsSize |      16,787kb | 16,781kb |   -7kb |  -0.04 |        BETTER | 16,774kb | 16,788kb | 
| jsStringsSize |         628kb |    628kb |    0kb |   0.02 |         WORSE |    628kb |    628kb | 
| jsOtherSize   |       8,399kb |  7,585kb | -813kb |  -9.68 |        BETTER |  6,722kb |  8,452kb | 
| otherSize     |         389kb |    388kb |    0kb |  -0.11 | INSIGNIFICANT |    374kb |    390kb | 

It also improves startup time on device, but the win is less significant, and there's a big memory hit:

|      Test     | Baseline Mean |   Mean   |    +/-   |    %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|---------:|-------:|--------------:|---------:|---------:|
| startupTime   |      12,388ms | 11,652ms |   -735ms |  -5.94 |        BETTER | 11,273ms | 12,481ms | 
| totalSize     |      22,523kb | 41,361kb | 18,838kb |  83.64 |         WORSE | 40,863kb | 41,739kb | 
| domSize       |          91kb |     78kb |    -13kb | -14.44 |        BETTER |     78kb |     78kb | 
| styleSize     |         291kb |    296kb |      5kb |   1.61 |         WORSE |    290kb |    305kb | 
| jsObjectsSize |      15,467kb | 30,981kb | 15,514kb | 100.30 |         WORSE | 30,968kb | 30,999kb | 
| jsStringsSize |         918kb |  1,322kb |    404kb |  43.99 |         WORSE |  1,322kb |  1,322kb | 
| jsOtherSize   |       5,579kb |  8,486kb |  2,907kb |  52.12 |         WORSE |  8,025kb |  8,788kb | 
| otherSize     |         177kb |    199kb |     22kb |  12.17 | INSIGNIFICANT |    166kb |    250kb | 

So this needs further investigation.

**Comment by mykmelez** [comment]
I made a few improvements, none of which I would have expected to impact memory usage, but a rerun on device shows the expected startup time win without any memory hit:

|      Test     | Baseline Mean |   Mean   |    +/-   |    %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|---------:|-------:|--------------:|---------:|---------:|
| startupTime   |      12,505ms | 11,427ms | -1,078ms |  -8.62 |        BETTER | 10,141ms | 12,364ms | 
| totalSize     |      22,293kb | 22,267kb |    -26kb |  -0.12 | INSIGNIFICANT | 21,677kb | 22,789kb | 
| domSize       |          91kb |     72kb |    -19kb | -20.70 |        BETTER |     72kb |     72kb | 
| styleSize     |         291kb |    296kb |      4kb |   1.47 | INSIGNIFICANT |    290kb |    307kb | 
| jsObjectsSize |      15,412kb | 15,417kb |      5kb |   0.03 | INSIGNIFICANT | 15,403kb | 15,437kb | 
| jsStringsSize |         635kb |    635kb |      0kb |   0.00 | INSIGNIFICANT |    635kb |    635kb | 
| jsOtherSize   |       5,686kb |  5,645kb |    -41kb |  -0.73 | INSIGNIFICANT |  4,954kb |  6,200kb | 
| otherSize     |         178kb |    203kb |     25kb |  13.80 | INSIGNIFICANT |    165kb |    284kb | 

I'm not sure why the results are different this time around.

I did notice that the pref *security.turn_off_all_security_so_that_viruses_can_take_over_this_computer* wasn't set on the device anymore, so I set it before doing the run. But it should have been set during the previous run too; otherwise, the benchmarking code shouldn't have done any memory benchmarking at all.

I also killed the app before reinstalling it after doing the baseline run. But that shouldn't have made a difference, unless somehow B2G reinstalls the app without restarting its process, and the existing process leaks the old window (which I sometimes see on desktop when reloading a page in the browser).

**Comment by mykmelez** [comment]
To double-check, I tested on another device, this one running a recent FxOS 3.0 nightly (the first device was running a recent FxOS 2.1 nightly). On the second device, the results are similar:

|      Test     | Baseline Mean |   Mean   |    +/-   |    %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|---------:|-------:|--------------:|---------:|---------:|
| startupTime   |      16,708ms | 14,922ms | -1,786ms | -10.69 |        BETTER | 13,673ms | 16,331ms | 
| totalSize     |      20,150kb | 20,067kb |    -83kb |  -0.41 | INSIGNIFICANT | 19,460kb | 20,313kb | 
| domSize       |          93kb |     74kb |    -19kb | -20.23 |        BETTER |     74kb |     74kb | 
| styleSize     |         287kb |    293kb |      6kb |   2.14 |         WORSE |    287kb |    299kb | 
| jsObjectsSize |      14,704kb | 14,661kb |    -43kb |  -0.30 |        BETTER | 14,644kb | 14,696kb | 
| jsStringsSize |         872kb |    578kb |   -293kb | -33.64 |        BETTER |    574kb |    604kb | 
| jsOtherSize   |       3,956kb |  4,184kb |    228kb |   5.77 | INSIGNIFICANT |  3,633kb |  4,326kb | 
| otherSize     |         238kb |    276kb |     38kb |  15.99 |         WORSE |    238kb |    314kb | 

So I think my original test results on device were a fluke, and this has no impact on memory.

**Comment by mykmelez** [comment]
(Since I was changing the code around it, and it's a trivial change, I went ahead and fixed #1305 in the process.)

File: Makefile [hunk 1]
```diff
@@ -106,6 +170,14 @@ bld/jsc.js: jsc.ts bld/j2me-jsc.js
 	@echo "Building J2ME JSC CLI"
 	node tools/tsc.js --sourcemap --target ES5 jsc.ts --out bld/jsc.js
 
+# Some scripts use ES6 features, so we have to specify ES6 as the in-language
+# (and ES5 as the out-language, since Closure doesn't recognize ES6 as a valid
+# out-language) in order for Closure to compile them, even though for now
+# we're optimizing "WHITESPACE_ONLY".
```
Comment (marco-c) [inline Makefile]: Do we have an issue opened to enable closure optimizations?

File: main.html.in [hunk 2]
```diff
@@ -10,82 +10,48 @@
...
+  <!-- Run `make aot` to generate these.  Load them separately, because adding
+     - classes.jar.js to main-all.js fails silently for unknown reasons. -->
   <script type="text/javascript" src="bld/classes.jar.js" defer></script>
   <!--<script type="text/javascript" src="bld/program.jar.js" defer></script>-->
   <!--<script type="text/javascript" src="bld/tests.jar.js" defer></script>-->
-  <script type="text/javascript" src="main.js" defer></script>
+
+  <!-- Load the rest of the scripts from the combined script main-all.js
+     - to improve startup and runtime performance.  See Makefile for the list
+     - of files that get combined into this one. -->
```
Comment (marco-c) [inline main.html.in]: Maybe in a followup we can investigate storing all the scripts in a singe file?

File: tools/package.sh [hunk 3]
```diff
@@ -8,12 +8,35 @@ PACKAGE_DIR="output"
 rm -rf $PACKAGE_DIR/
 mkdir $PACKAGE_DIR
 mkdir $PACKAGE_DIR/bld
+mkdir $PACKAGE_DIR/config
+mkdir $PACKAGE_DIR/libs
+mkdir $PACKAGE_DIR/polyfill
 
 # setup the root
-cp *.js *.html *.webapp $PACKAGE_DIR/.
+# TODO: only copy benchmark.js/gc.html/gc.js if BENCHMARK=1.
```
Comment (marco-c) [inline tools/package.sh]: Can you file an issue for this and the other TODO?

File: Makefile (hunk 1 above)
Comment (mykmelez) [inline Makefile]: We didn't, so I opened https://github.com/mozilla/j2me.js/issues/1339.

File: main.html.in (hunk 2 above)
Comment (mykmelez) [inline main.html.in]: Good idea, I filed https://github.com/mozilla/j2me.js/issues/1338 on it.

File: tools/package.sh (hunk 3 above)
Comment (mykmelez) [inline tools/package.sh]: Yup, filed as https://github.com/mozilla/j2me.js/issues/1340.

**Pull request merged**